feat(@schematics/angular): add option to setup new workspace or appli…

aparzi opened the pull request:
Introduces option `--experimental-zoneless` to setup workspace or application as zoneless mode.

## PR Checklist

Please check to confirm your PR fulfills the following requirements:

<!-- Please check all that apply using "x". -->

- [X] The commit message follows our guidelines: https://github.com/angular/angular-cli/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Closes #28370
## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

aparzi commented:
Hi @alan-agius4,
Based on the email exchange we had, I created the draft PR, to check the reason why it does not recognize the experimental-zoneless option and possibly give me some feedback and suggestions.

Thanks.

alan-agius4 commented on packages/schematics/angular/application/index.ts:
```diff
@@ -132,6 +136,9 @@ function addDependenciesToPackageJson(options: ApplicationOptions) {
       context.addTask(new NodePackageInstallTask());
     }
+    if (options.experimentalZoneless) {
+      removePackageJsonDependency(host, 'zone.js');
```
We should avoid removing `zone.js` from the `package.json`, as this could break other projects within the monorepo that depend on it.

alan-agius4 commented on packages/schematics/angular/ng-new/schema.json:
```diff
@@ -138,6 +138,11 @@
       "x-user-analytics": "ep.ng_ssr"
+    },
+    "experimentalZoneless": {
+      "description": "Create an application without dependencies related to zone.js (zoneless)",
```
```suggestion
      "description": "Create an application that does not utilize zone.js."
```

alan-agius4 reviewed (commented):
Thanks for this, overall this looks good. 

Would be great if you can also add some UTs.

aparzi commented on packages/schematics/angular/application/index.ts:
```diff
@@ -132,6 +136,9 @@ function addDependenciesToPackageJson(options: ApplicationOptions) {
     if (!options.skipInstall) {
       context.addTask(new NodePackageInstallTask());
     }
+    if (options.experimentalZoneless) {
```
Are you referring to the monorepo projects present in the CLI?

aparzi commented:
Hi @alan-agius4,
I updated the PR according to the requested changes. I added a test case, the only thing running the command `yarn bazel test //packages/schematics/angular:angular_test` seems to give the following error: 
```SyntaxError: Unexpected token ')'.```
This however even without writing new test cases, seems to be related to the runSchematic method used in the .spec files

alan-agius4 commented:
> Hi @alan-agius4, I updated the PR according to the requested changes. I added a test case, the only thing running the command `yarn bazel test //packages/schematics/angular:angular_test` seems to give the following error: `SyntaxError: Unexpected token ')'.` This however even without writing new test cases, seems to be related to the runSchematic method used in the .spec files

The problem is that in the templates, you have missing some `<%`.

alan-agius4 commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 The MIT License
```
Incorrect rebase?

alan-agius4 commented on packages/schematics/angular/application/files/standalone-files/src/app/app.config.ts.template:
```diff
@@ -4,5 +4,7 @@ import { provideRouter } from '@angular/router';
 export const appConfig: ApplicationConfig = {
-  providers: [provideZoneChangeDetection({ eventCoalescing: true })<% if (routing) { %>, provideRouter(routes)<% } %>]
+  providers: [
+  <% if(experimentalZoneless) { %>providers: [provideExperimentalZonelessChangeDetection()],} %>
```
```suggestion
  <% if(experimentalZoneless) { %>providers: [provideExperimentalZonelessChangeDetection()],<% } %>
```

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -4,6 +4,7 @@
 import { AppModule } from './app/app.module';
 
 platformBrowserDynamic().bootstrapModule(AppModule, {
+  <% if(experimentalZoneless) { %>providers: [provideExperimentalZonelessChangeDetection()],} %>
```
```suggestion
  <% if(experimentalZoneless) { %>providers: [provideExperimentalZonelessChangeDetection()],<% } %>
```

alan-agius4 commented on packages/schematics/angular/ng-new/index_spec.ts:
```diff
@@ -95,12 +95,21 @@ describe('Ng New Schematic', () => {
   });
 
-  it('should add packageManager option in angular.json', async () => {
+  fit('should add packageManager option in angular.json', async () => {
```
```suggestion
  it('should add packageManager option in angular.json', async () => {
```

aparzi commented:
Hi @alan-agius4,
I will adjust the PR as soon as possible. Can you also tell me why when I run the command `ng new [angular_project_name] --experimental-zoneless` it gives me an error saying that the option is not recognized?

alan-agius4 commented:
> Can you also tell me why when I run the command ng new [angular_project_name] --experimental-zoneless it gives me an error saying that the option is not recognized?

Likely because you do globally you do not have the local copy

aparzi commented:
Hi @alan-agius4,
I updated the PR according to the requested changes and some improvements and added the test cases. I'm waiting for news.

Thanks

alan-agius4 commented on packages/schematics/angular/workspace/schema.json:
```diff
@@ -41,6 +41,11 @@
       "description": "The package manager used to install dependencies.",
       "type": "string",
       "enum": ["npm", "yarn", "pnpm", "cnpm", "bun"]
+    },
+    "experimentalZoneless": {
+      "description": "Create an application without dependencies related to zone.js (zoneless)",
+      "type": "boolean",
+      "default": false
```
Unused option

```suggestion
```

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -4,6 +4,7 @@
 import { AppModule } from './app/app.module';
 
 platformBrowserDynamic().bootstrapModule(AppModule, {
+  <% if(experimentalZoneless) { %>providers: [provideExperimentalZonelessChangeDetection()],<% } %>
   ngZoneEventCoalescing: true<% if(!!viewEncapsulation) { %>,
```
It looks like we it's zoneless `ngZoneEventCoalescing` is still added.

```suggestion
<% if(experimentalZoneless) { %>providers: [provideExperimentalZonelessChangeDetection()]<% } else { %>ngZoneEventCoalescing: true<% if(!!viewEncapsulation) { %>,
```

alan-agius4 commented on packages/schematics/angular/application/files/standalone-files/src/app/app.config.ts.template:
```diff
@@ -4,5 +4,8 @@ import { provideRouter } from '@angular/router';
-  providers: [provideZoneChangeDetection({ eventCoalescing: true })<% if (routing) { %>, provideRouter(routes)<% } %>]
+  providers: [
+  <% if(experimentalZoneless) { %>provideExperimentalZonelessChangeDetection(),<% } %>
+  provideZoneChangeDetection({ eventCoalescing: true })<% if (routing) { %>, provideRouter(routes)<% } %>
```
`provideZoneChangeDetection` is still added when using zoneless. This should not be the case.

alan-agius4 commented on packages/schematics/angular/ng-new/index_spec.ts:
```diff
@@ -103,4 +103,22 @@ describe('Ng New Schematic', () => {
+    expect(fileContent).toContain('provideExperimentalZonelessChangeDetection()');
+  });
+
+  it('should experimentalZoneless is set to false', async () => {
```
`should not add 'provideExperimentalZonelessChangeDetection()' when 'experimentalZoneless' is false`

alan-agius4 commented on packages/schematics/angular/ng-new/index_spec.ts:
```diff
@@ -103,4 +103,22 @@ describe('Ng New Schematic', () => {
     expect(cli.packageManager).toBe('npm');
   });
+
+  it('should experimentalZoneless is set to true and add provideExperimentalZonelessChangeDetection in app.config', async () => {
```
`should add 'provideExperimentalZonelessChangeDetection()' when 'experimentalZoneless' is true`

alan-agius4 commented on packages/schematics/angular/ng-new/index_spec.ts:
```diff
@@ -103,4 +103,22 @@ describe('Ng New Schematic', () => {
+    });
+    const fileContent = tree.readContent('/bar/src/app/app.config.ts');
+    expect(fileContent).not.toMatch('provideExperimentalZonelessChangeDetection()');
+  });
```
Can you also please add a test to check that `provideZoneChangeDetection` is not added when it's in zoneless?

alan-agius4 commented on packages/schematics/angular/ng-new/index_spec.ts:
```diff
@@ -103,4 +103,22 @@ describe('Ng New Schematic', () => {
+      experimentalZoneless: false,
+    });
+    const fileContent = tree.readContent('/bar/src/app/app.config.ts');
+    expect(fileContent).not.toMatch('provideExperimentalZonelessChangeDetection()');
```
```suggestion
    expect(fileContent).not.toContain('provideExperimentalZonelessChangeDetection()');
```

alan-agius4 commented on packages/schematics/angular/ng-new/index_spec.ts:
```diff
@@ -103,4 +103,22 @@ describe('Ng New Schematic', () => {
     const { cli } = JSON.parse(tree.readContent('/bar/angular.json'));
     expect(cli.packageManager).toBe('npm');
   });
+
```
These tests are more appropriately placed in the application schematic spec file, as they test code within the application schematic. None of this code tested code is in `ng-new`.

aparzi replied:
Hi @alan-agius4, do you suggest me to move the test cases inside application? In the index.spec file?

alan-agius4 replied:
Yes

aparzi commented:
@alan-agius4 I believe all request changes done.

alan-agius4 commented:
Can you also please add an e2e to make sure that the code generated is valid?

Something like the following;

**tests/legacy-cli/e2e/tests/generate/application/application-zoneless.ts**
```ts
import { ng } from '../../../utils/process';
import { useCIChrome } from '../../../utils/project';

export default async function () {
  await ng('generate', 'application', 'app2', '--experimental-zoneless');
  await useCIChrome('app2', 'projects/app2');
  await ng('test', 'app2', '--watch=false');
  await ng('build', 'app2');
}
```

aparzi commented:
Hi @pweyrich @alan-agius4 
I updated the PR with the requested changes. Let me know if everything is ok.

Many thanks.

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -1,10 +1,16 @@
-})
+const bootstrapOptions = {
+    <% if(!experimentalZoneless) { %>ngZoneEventCoalescing: true<% } else { %>
+    const { provideExperimentalZonelessChangeDetection } = await import('@angular/core');
```
Please revert this to use top level import instead of a top level await.

alan-agius4 commented on tests/legacy-cli/e2e/tests/generate/application/application-zoneless.ts:
```diff
@@ -0,0 +1,9 @@
+export default async function () {
+  await ng('generate', 'application', 'app2', '--experimental-zoneless');
+  await useCIChrome('app2', 'projects/app2');
+  await ng('test', 'app2', '--watch=false');
```
Looks like `ng test` is failing because `provideExperimentalZonelessChangeDetection` need to be added also added in `app.component.spec.ts`

aparzi replied:
Okay @alan-agius4

aparzi replied:
@alan-agius4 In each app.component.spec.ts that are in the assets folder do I need to add provideExperimentalZonelessChangeDetection() ?

alan-agius4 replied:
Yes, in https://github.com/angular/angular-cli/blob/c5bad578f9367fc705e5a3a6afcd89f944cc3966/packages/schematics/angular/application/files/standalone-files/src/app/app.component.spec.ts.template#L6 and https://github.com/angular/angular-cli/blob/c5bad578f9367fc705e5a3a6afcd89f944cc3966/packages/schematics/angular/application/files/module-files/src/app/app.component.spec.ts.template#L7

See: https://angular.dev/guide/experimental/zoneless#using-zoneless-in-testbed

aparzi commented:
Hi @alan-agius4,
I updated this PR. Let me know if everything is ok.

Many thanks.

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/app/app.component.spec.ts.template:
```diff
@@ -1,6 +1,7 @@
 import { TestBed } from '@angular/core/testing';<% if (routing) { %>
 import { RouterModule } from '@angular/router';<% } %>
 import { AppComponent } from './app.component';
+<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection from '@angular/core'; <% } %>
```
Nit move the below to the first line, also it has a missing closing curly.

```suggestion
<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection } from '@angular/core'; <% } %>
```

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -1,10 +1,16 @@
-  ngZoneEventCoalescing: true<% if(!!viewEncapsulation) { %>,
-  defaultEncapsulation: ViewEncapsulation.<%= viewEncapsulation %><% } %>
-})
+platformBrowserDynamic().bootstrapModule(AppModule, bootstrapOptions)
```
Just noticed but, `providers` is not valid in this context, it needs to be added to the ngModule. 

https://github.com/angular/angular-cli/blob/d6dec2309f73417111b81e44f0a393d805f4739c/packages/schematics/angular/application/files/module-files/src/app/app.module.ts.template

```suggestion
platformBrowserDynamic().bootstrapModule(AppModule, {
  <% if(!experimentalZoneless) { %>ngZoneEventCoalescing: true, %><% if(!!viewEncapsulation) { %>
  defaultEncapsulation: ViewEncapsulation.<%= viewEncapsulation %><% } %>
})
```


See: https://angular.dev/guide/experimental/zoneless

alan-agius4 replied:
Looks like this has not been done. IE: the change in `app.module.ts.template`

aparzi replied:
it seems to be ok or i received it wrong how to do the correction.

alan-agius4 replied:
Unless, `provideExperimentalZonelessChangeDetection` is added to app.module that ngModule based app will be configured to use zone.js, even when using `—experimental-zoneless`.

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -1,10 +1,16 @@
-<% if(!!viewEncapsulation) { %>import { ViewEncapsulation } from '@angular/core';
-<% }%>import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
-
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 import { AppModule } from './app/app.module';
+<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection from '@angular/core'; <% } %>
+
+const bootstrapOptions = {
+    <% if(!experimentalZoneless) { %>ngZoneEventCoalescing: true<% } else { %>
+    providers: [<% if(experimentalZoneless) { %>provideExperimentalZonelessChangeDetection()<% } %>];
+    <% } %>
+    <% if(!!viewEncapsulation) { %>
+      const { ViewEncapsulation } = await import('@angular/core');
+      defaultEncapsulation: ViewEncapsulation.<%= viewEncapsulation %>;
+    <% } %>
+};
```
```suggestion
<% if(!!viewEncapsulation) { %>import { ViewEncapsulation } from '@angular/core';
<% }%>import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
```

alan-agius4 commented on packages/schematics/angular/application/files/standalone-files/src/app/app.component.spec.ts.template:
```diff
@@ -1,10 +1,12 @@
 import { TestBed } from '@angular/core/testing';
 import { AppComponent } from './app.component';
+<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection from '@angular/core'; <% } %>
```
Nit: move this as first import. Also there is a missing `}`

```suggestion
<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection } from '@angular/core'; <% } %>
```

alan-agius4 reviewed (changes requested):
thanks, see above.

aparzi commented:
Hi @alan-agius4,
I have updated the PR with the requested changes. Sorry for the inconvenience.

Many thanks.

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/app/app.component.spec.ts.template:
```diff
@@ -1,6 +1,7 @@
 import { TestBed } from '@angular/core/testing';<% if (routing) { %>
 import { RouterModule } from '@angular/router';<% } %>
 import { AppComponent } from './app.component';
+<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection } from '@angular/core'; <% } %>
```
As per previous comment, can you move this to first line please?

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -1,10 +1,9 @@
+<% if(!!viewEncapsulation) { %>import { ViewEncapsulation } from '@angular/core';<% }%>
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 import { AppModule } from './app/app.module';
+<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection from '@angular/core'; <% } %>
```
```suggestion

```

alan-agius4 replied:
This is unused

```suggestion
```

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -1,10 +1,9 @@
-<% if(!!viewEncapsulation) { %>import { ViewEncapsulation } from '@angular/core';
-<% }%>import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
-
+<% if(!!viewEncapsulation) { %>import { ViewEncapsulation } from '@angular/core';<% }%>
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
```
```suggestion
<% if(!!viewEncapsulation) { %>import { ViewEncapsulation } from '@angular/core';
<% }%>import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
```

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/app/app.component.spec.ts.template:
```diff
@@ -1,6 +1,7 @@
 import { TestBed } from '@angular/core/testing';<% if (routing) { %>
```
```suggestion
<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection } from '@angular/core'; <% } %>import { TestBed } from '@angular/core/testing';<% if (routing) { %>
```

alan-agius4 commented on packages/schematics/angular/application/files/standalone-files/src/app/app.component.spec.ts.template:
```diff
@@ -1,10 +1,12 @@
 import { TestBed } from '@angular/core/testing';
 import { AppComponent } from './app.component';
+<% if(experimentalZoneless) { %> import { provideExperimentalZonelessChangeDetection } from '@angular/core'; <% } %>
```
Same as prev, please move this import in Line 1.

alan-agius4 commented on packages/schematics/angular/application/index_spec.ts:
```diff
@@ -698,5 +698,48 @@ describe('Application Schematic', () => {
+        workspaceTree,
+      );
+      const path = '/projects/foo/src/app/app.config.ts';
+      console.log('tree => ', tree);
```
```suggestion
```

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -1,10 +1,9 @@
 
 platformBrowserDynamic().bootstrapModule(AppModule, {
-  ngZoneEventCoalescing: true<% if(!!viewEncapsulation) { %>,
+  <% if(!experimentalZoneless) { %>ngZoneEventCoalescing: true, %><% } %><% if(!!viewEncapsulation) { %>
```
```suggestion
  <% if(!experimentalZoneless) { %>ngZoneEventCoalescing: true,<% } %><% if(!!viewEncapsulation) { %>
```

alan-agius4 reviewed (changes requested):
Thanks, there are still a few issues.

aparzi commented:
> Thanks, there are still a few issues.

Ok thanks, I have updated the PR with the requested changes.

alan-agius4 commented on packages/schematics/angular/application/files/module-files/src/main.ts.template:
```diff
@@ -1,10 +1,8 @@
+  <% if(!experimentalZoneless) { %>ngZoneEventCoalescing: true,<% } %><% if(!!viewEncapsulation) { %>
   defaultEncapsulation: ViewEncapsulation.<%= viewEncapsulation %><% } %>
 })
-  .catch(err => console.error(err));
```
This has been removed by mistake. Kindly restore.

alan-agius4 commented on tests/legacy-cli/e2e/tests/generate/application/application-zoneless.ts:
```diff
@@ -0,0 +1,9 @@
+import { useCIChrome } from '../../../utils/project';
+
+export default async function () {
+  await ng('generate', 'application', 'app2', '--experimental-zoneless');
+  await useCIChrome('app2', 'projects/app2');
+  await ng('test', 'app2', '--watch=false');
+  await ng('build', 'app2');
```
Maybe it would be better to also add a test for non standalone. 


```suggestion
await ng('generate', 'app', 'standalone', '--experimental-zoneless', '--standalone');
await useCIChrome('standalone', 'projects/standalone');
await ng('test', 'standalone', '--watch=false');
await ng('build', 'standalone');

await ng('generate', 'app', 'ngmodules', '--experimental-zoneless', '--no-standalone', '--skip-install');
await useCIChrome('ngmodules', 'projects/ngmodules');
await ng('test', 'ngmodules', '--watch=false');
await ng('build', 'ngmodules');
```

alan-agius4 reviewed (commented):
Thanks again for this, almost there.

aparzi commented:
> Thanks again for this, almost there.

I think I did everything. I just have doubts about e2e if they pass, I tried to follow what you suggested.

Thanks again.

aparzi commented:
@alan-agius4 
I updated the PR adding the requested change in app.module. To be safe I also added two more test cases, after this change.

Thanks I'm waiting

alan-agius4 reviewed (approved):
LGTM, thanks
